packages/block-library/src/template-part/edit/template-part-previews.js

```diff
@@ -0,0 +1,209 @@
+				{ children }
+			</div>
+		</>
+	);
+}
+
+function TemplatePartsByTheme( { templateParts, selectTemplate } ) {
+	const templatePartsByTheme = useMemo( () => {
+		return Object.values( groupBy( templateParts, 'meta.theme' ) );
+	}, [ templateParts ] );
+	const currentShownTPs = useAsyncList( templateParts );
+
+	return (
+		<>
+			{ templatePartsByTheme.length &&
+				templatePartsByTheme.map( ( templatePartList ) => (
+					<PanelGroup
+						key={ templatePartList[ 0 ].meta.theme }
+						title={ templatePartList[ 0 ].meta.theme }
+					>
+						{ templatePartList.map( ( templatePart ) => {
+							return currentShownTPs.includes( templatePart ) ? (
+								<TemplatePartItem
+									key={ templatePart.id }
+									templatePart={ templatePart }
+									selectTemplate={ selectTemplate }
+								/>
+							) : (
+								<TemplatePartPlaceholder
+									key={ templatePart.id }
+								/>
+							);
+						} ) }
+					</PanelGroup>
+				) ) }
+		</>
+	);
+}
+
+function TemplatePartSearchResults( {
+	templateParts,
+	selectTemplate,
+	filterValue,
+} ) {
+	const filteredTPs = useMemo( () => {
+		// Filter based on value.
+		const lowerFilterValue = filterValue.toLowerCase();
+		const searchResults = templateParts.filter(
+			( { slug, meta: { theme } } ) =>
+				slug.toLowerCase().includes( lowerFilterValue ) ||
+				theme.toLowerCase().includes( lowerFilterValue )
+		);
+		// Order based on value location.
+		searchResults.sort( ( a, b ) => {
+			// First prioritize index found in slug.
+			const indexInSlugA = a.slug
+				.toLowerCase()
+				.indexOf( lowerFilterValue );
+			const indexInSlugB = b.slug
+				.toLowerCase()
+				.indexOf( lowerFilterValue );
+			if ( indexInSlugA !== -1 && indexInSlugB !== -1 ) {
+				return indexInSlugA - indexInSlugB;
+			} else if ( indexInSlugA !== -1 ) {
+				return -1;
+			} else if ( indexInSlugB !== -1 ) {
+				return 1;
+			}
+			// Second prioritize index found in theme.
+			return (
+				a.meta.theme.toLowerCase().indexOf( lowerFilterValue ) -
+				b.meta.theme.toLowerCase().indexOf( lowerFilterValue )
+			);
+		} );
+		return searchResults;
+	}, [ filterValue, templateParts ] );
+
+	const currentShownTPs = useAsyncList( filteredTPs );
+
+	return (
+		<>
+			{ filteredTPs.map( ( templatePart ) => (
+				<PanelGroup
+					key={ templatePart.id }
+					title={ templatePart.meta.theme }
+				>
+					{ currentShownTPs.includes( templatePart ) ? (
+						<TemplatePartItem
+							key={ templatePart.id }
+							templatePart={ templatePart }
+							selectTemplate={ selectTemplate }
+						/>
+					) : (
+						<TemplatePartPlaceholder key={ templatePart.id } />
+					) }
+				</PanelGroup>
+			) ) }
+		</>
+	);
+}
+
+export default function TemplateParts( { selectTemplate, filterValue } ) {
+	const templateParts = useSelect( ( select ) => {
+		return select( 'core' ).getEntityRecords(
+			'postType',
+			'wp_template_part',
+			{
+				status: [ 'publish', 'auto-draft' ],
+			}
+		);
```

Comment:
We shouldn't show the non-customized versions of customized template parts.

We also shouldn't show all the potential temporary auto-drafts.

We need to use the `resolved` query parameter for fetching all template parts for the current theme, and we need to merge that with a fetch for all published template parts.

Reply:
> We need to use the `resolved` query parameter for fetching all template parts for the current theme, and we need to merge that with a fetch for all published template parts.

Ok!  I think I am following here.  So in merging these 2 results, we would get rid of items in the `resolved` results that have customized counterparts in the `published` ones?

Reply:
The `resolved` result will also contain the customized versions of the resolved template parts.

So, we would actually deduplicate them when merging.

Reply:
>I did not end up with a "customized" and "un-customized" version though, just the customized one both in the 'resolved' and 'published' results.

That's expected. What did you expect exactly?

>resolved is not supplying all template parts from the theme. 

That was fixed on master earlier today. Try rebasing.

>Oddly, it seems like resolved: true and resolved: false give the same results when used for templates or template parts.

The code pathways are very different. It might be a coincidence with your configuration.

Reply:
> That's expected. What did you expect exactly?

Oh ok, good!  From earlier in the discussion I was expecting to have both versions and having to take additional steps to 'hide' the non-customized one.

> That was fixed on master earlier today. Try rebasing.

Oh awesome!  I did rebase earlier today but probably just a bit too early.

Reply:
You may have two copies of the customized version.

Let me know if the rebase fixes it.

Reply:
It doesn't seem to fix it, I'm still having the same issue.

Using 'twentynineteen-blocks', the `resolved` query is giving me the footer for this theme.  The header for this theme doesn't show up in either of the queries, but does appear under the 'auto-draft' if i add that.  🤔

Reply:
Do you see it listed in the dropdown of the site editor? That uses the same query.

Reply:
That dropdown only has the demo template header (theme name 'twentytwenty' is the demo header iirc?).  Even with demo templates turned off, its probably still there after being previously saved with a customization.  But that dropdown doesn't show the footer or the header that are supplied by the theme.

I didn't mention before, my `resolved` query also returns that demo header for 'twentytwenty' (I figured that was expected?) along with the footer from twentytwenty-blocks.  But that twentytwenty-blocks header doesn't show up from that.

Reply:
`resolved` returns all template parts that are reference by all of your resolved templates. Try wiping your environment. I can't reproduce this with my setup.

Reply:
Ok, wiping the environment fixes things for the most part.  Both theme supplied template parts are appearing under the resolved query as expected (regardless of having demo templates enabled/disabled).

Without demo templates both theme supplied header and footer appear in the dropdown.

With the demo templates applied, only the header from the theme is in the dropdown. I guess this is expected as the demo template doesn't have a footer.  It's interesting that its using the header from the theme inside the demo template.  But being that its a demo experiment I don't think its too concerning.

Reply:
> resolved returns all template parts that are reference by all of your resolved templates. 

Would this mean that if I customized/saved a theme supplied template to use a different header template part that the theme supplied header would no longer be resolvable?

Reply:
>It's interesting that its using the header from the theme inside the demo template. But being that its a demo experiment I don't think its too concerning.

Is the theme name the same between the two? That shouldn't be happening.

>Would this mean that if I customized/saved a theme supplied template to use a different header template part that the theme supplied header would no longer be resolvable?

If it's not used by any templates, yes.

Good catch, we need to fix that. I think we need to modify the `theme` query to create auto drafts for the template part block files and include them along with any customized published ones.

That way, we can do a `theme` query for the active theme and combine that with the published/customized query.

Does that make sense?

Reply:
> Is the theme name the same between the two? That shouldn't be happening.

The header being used by the demo template is from / theme field is `twentytwenty-blocks`.

I encountered this by:  wiping the environment -> enabled 2020-blocks theme -> enabled FSE experiment (no demos) -> visited site editor -> enabled demo templates -> went back to site editor.  

> Does that make sense?

Kind of.  We would set up auto-drafts to be created for template parts supplied by themes.  Then we could do a third `getEntityRecords` call here to query the theme name and allow auto-draft as an acceptable status in that case?  So the final list here would be a unique combination of results from the published, resolved, and theme/auto-draft queries?

Reply:
>The header being used by the demo template is from / theme field is twentytwenty-blocks.

That's a bug.

>So the final list here would be a unique combination of results from the published, resolved, and theme/auto-draft queries?

The last two are a superset of resolved so we can drop that.

Reply:
> The last two are a superset of resolved so we can drop that.

Ah, yeah that makes sense.

> I think we need to modify the theme query to create auto drafts for the template part block files

Is this something that needs to be taken care of before this PR lands?  Or can we follow-up and update once thats in place?

Reply:
>Is this something that needs to be taken care of before this PR lands? Or can we follow-up and update once thats in place?

Whatever you think is best.

It should be a quick addition to the REST filter for template parts.

Reply:
> Whatever you think is best.

Handling separately may be good,  Im still a bit uncertain what/how to do for that query.  I think this has been updated for all other provided feedback so far. 

> It should be a quick addition to the REST filter for template parts.

Does requesting the theme and including auto-draft in the status field not give us this?
ex
```
			{
				theme: 'theme-name',
				status: [ 'publish', 'auto-draft' ],
				per_page: -1,
			}
```
would that be a substitute for the 'resolved' query that also adds those in?

I assume by the theme query / rest filter you are referring to:
https://github.com/WordPress/gutenberg/blob/b83f94845501378dfd78f1a10db785e05dee2d87/lib/template-parts.php#L209-L217

Did we just want to add the post_status args there?  or is there something more specific we need to do to 'create' the auto-drafts as you mention?

Reply:
>Does requesting the theme and including auto-draft in the status field not give us this?

No, because nothing guarantees the auto-drafts have been created.

>would that be a substitute for the 'resolved' query that also adds those in?

Yes, see https://github.com/WordPress/gutenberg/pull/22760#discussion_r439151028.

>I assume by the theme query / rest filter you are referring to:
>Did we just want to add the post_status args there? or is there something more specific we need to do to 'create' the auto-drafts as you mention?

If `theme === currentTheme`, we need to pass all its template part files through `create_auto_draft_for_template_part_block`.

Reply:
It will probably be best to handle that in a separate PR.  It probably is very simple, maybe even slapping my in the face, but Im not completely seeing it and am a bit distracted with other items at the immediate moment.  Il see if I can look at it a bit more later, but separate to this PR.

> No, because nothing guarantees the auto-drafts have been created.
> Yes, see #22760 (comment).

Sorry, I was poorly asking the same question twice.  So yes, but we need to ensure the auto-drafts are saved.   I guess I was a bit confused as I was finding auto-drafts saved for my theme supplied template parts already.

Reply:
@epiqueras  - 
> we need to pass all its template part files through `create_auto_draft_for_template_part_block`.

whats the best way of getting those template part files for the corresponding theme in this function?  Im still a bit lost on how to implement this. 

 I see `gutenberg_find_template_post_and_parts` but that returns their IDs (which the template parts we are concerned with wouldn't have yet), similarly Im not sure what I would pass in as the template_types.  Is there another useful function to retrieve those theme supplied template parts that we need to create auto drafts of?

Reply:
actually I think @ockham pointed me in the right direction here.  I have an idea now? 😆

Reply:
That function creates the auto drafts, but yeah, it needs a template type.

You need to mirror what the code there is doing for finding template files, but for template parts. You could even abstract it into a helper, `get_template_files`, and `get_template_part_files`.

Reply:
@ockham pointed out that `gutenberg_find_template_post_and_parts` already handles locating the block template files and creating the template part auto drafts.  So im wondering if something like this will work: https://github.com/WordPress/gutenberg/pull/23254

we can continue discussion over there.